### ISSUE TITLE
RFC: Refactor library with improvements but breaking changes

### ISSUE THREAD

**Issue by cristiangraz** [blacklightcms/recurly#116]
This library was initially one of my earlier Go projects when I was new to the language. As the library has grown in usage, many PRs have shown design issues surfacing and there have been several great suggestions from the community.

I have started working on a refactor for the library that will address all of the issues up until this point, and as a result be a breaking change.

I'd like to solicit some feedback from the community on this. 

## Goals
- Remove triple response fields and simplify error handling
- Ensure all custom error types implement the `error` interface
- Revisit all XML marshalers/unmarshalers
- Cleanup test suite
- Cleanup all deprecated functionality
- Improve method names
- Standardize and simplify pagination
- Reduce extra structs (e.g. NewSubscription just to Subscription with cleanup in method. See Billing.Update)
- Support the `context` package

## Release Plan
The goal would be to discuss some of the ideas here, then push up a `refactor` branch on an open PR to allow time for review as well as time for anyone in the community to try it out and report bugs.

We would tag a release of the current master branch as something like `0.1`, then when this change is ready tag it as a major version update because it will be a breaking change.

We have a pretty large Recurly integration and I anticipate the changes as shown here would take 2-3 hours of total time to update, but would drastically simplify much of our code (particularly around error handling).

## Specifics
Here are the specifics of each of the goals.

### Remove triple response fields and simplify error handling (https://github.com/blacklightcms/recurly/issues/47)
This is the biggest issue I see come up with the library. We always return a `*recurly.Response`, and in addition to checking that `err != nil` you also have to check the response, which leads many to have to create their own error wrapper functions.

The response is helpful to know what status code was returned, and to access transaction errors, failed transaction, or validation errors.

We can simplify by making `*recurly.Response` unexported and introducing a few new error types:

```go
// RateLimitError occurs when Recurly returns a 429 Too Many Requests error.
type RateLimitError struct {
	Response *http.Response

	Rate Rate // Rate specifies the last known rate limit for the client
}

// TransactionFailedError is returned when a transaction fails.
type TransactionFailedError struct {
	Response *http.Response

	// Transaction holds the failed transaction (if any).
	Transaction *Transaction

	// TransactionError holds the transaction error. Recurly sometimes returns this
	// as an independent block, sometimes inside of a transaction,
	// and sometimes as both. This field will always be set regardless of
	// where the TransactionError appears.
	TransactionError TransactionError
}

// ClientError occurs when Recurly returns 400-499 status code.
// There are two known exceptions to this:
// 1) 429 Too Many Requests. See RateLimitError
// 2) 422 Unprocessable Entity with a failed transaction. See TransactionFailedError.
type ClientError struct {
	Response *http.Response

	// ValidationErrors holds an array of validation errors if any occurred.
	ValidationErrors []ValidationError
}

// ServerError occurs when recurly returns 500-599 status code.
type ServerError struct {
	Response *http.Response
}
```

Once that is in place, we can avoid returning a response at all and return either `error` or `(*Subscription, error)` (for example).

`RateLimitError`, `ServerError` can generally be treated as normal errors unless you have a specific need to type assert and handle explicitly.

`TransactionFailedError` can be used to access failed transactions, and `ClientError` can be used to look for validation errors. An example might look like this:

```go
purchase, err := c.Purchases.Create(...)
if e, ok := err.(*recurly.TransactionFailedError); ok {
	// access to e.TransactionError, and if set, e.Transaction
        // This only needs to be checked on API calls that process a transaction
} else if e, ok := err.(*recurly.ClientError); ok {
        // validation errors can be explicitly checked here if desired, or most likely this can be 
        // caught by the next block depending on use case
} else if err != nil {
        // all errors
}
```
The benefit is that a library user now only has to check if `err != nil`, and can opt-into type assertions of errors either conditionally (e.g. transaction errors) or to have more control should they choose. The current implementation forces users to have complex checks (checking errors and the response object).

There is a great explanation of the issue [here](https://github.com/blacklightcms/recurly/issues/47) by @MOZGIII

> Note I would only use it (helper) when I want to handle recurly-specific errors differently than any other one, which I don't need to do currently in many cases. This variant of the API really improves the cases where you don't care what kind of error occured, as you only need to do one check. Plus it's less error-prone, since the function signature is simpler and it is therefor more easy to work with correctly.

### Ensure all custom error types implement the `error` interface (https://github.com/blacklightcms/recurly/pull/78)
All of the errors above implement the error interface so they can be used as normal Go errors.

### Revisit all XML marshalers/unmarshalers
We have a lot of custom XML marshalers and unmarshalers, but there is a lot of duplication and/or extra code that could be eliminated.

e.g. Compare this:
https://github.com/blacklightcms/recurly/blob/567d9241157b50a6dd378787bacd000118ae99f0/accounts.go#L54-L114

to:
```go
func (a Address) MarshalXML(e *xml.Encoder, start xml.StartElement) error {
	if a.Address == "" && a.Address2 == "" && a.City == "" && a.State == "" && a.Zip == "" && a.Country == "" && a.Phone == "" {
		return nil
	}
	type alias Address
	return e.Encode(alias(a))
}
```

### Cleanup test suite
Moving xml into `testdata/account.xml` (for example) and providing functions in the test package to return a matching type. A test would look like this:

```go
func TestAccounts_Balance(t *testing.T) {
	setup()
	defer teardown()

	var invoked bool
	mux.HandleFunc("/v2/accounts/1/balance", func(w http.ResponseWriter, r *http.Request) {
		invoked = true
		if r.Method != "GET" {
			t.Fatalf("unexpected method: %s", r.Method)
		}
		w.WriteHeader(http.StatusOK)
		w.Write(MustOpenFile("account_balance.xml"))
	})

	if balance, err := client.Accounts.Balance("1"); !invoked {
		t.Fatal("expected fn invocation")
	} else if err != nil {
		t.Fatal(err)
	} else if diff := cmp.Diff(balance, NewTestAccountBalance()); diff != "" {
		t.Fatal(diff)
	}
}
```

### Cleanup all deprecated functionality
Creating transactions is deprecated, Recurly is pushing users to the purchases API.

Some deprecated fields still exist on many of our structs (coupons has a few)

### Improve method names
A few of the method names are overly verbose. So far I plan on changing `Accounts.LookupAccountBalance()` to `Accounts.Balance()`

### Standardize and simplify pagination
Pagination is really ugly in the current library (see README for example usage).

My thought is to provide a callback that is auto-populated on each instance, with a bool to stop.

```List(Params, func([]Account) (stop bool)) error```

The other option I explored was returning a `Pager` type to all paginated methods, we just end up with an extra return value.

```List(Params) ([]Account, Pager, error)```

### Reduce extra structs (e.g. NewSubscription just to Subscription with cleanup in method. See Billing.Update)
This comes up often. We have a `NewSubscription` struct. `UpdateSubscription`. `PurchaseSubscription`.

What I am considering is having a single `Subscription` struct, but in the API functions tucking away the implementation details for what fields are sent without exposing the entire struct. For example, `Subscription.Create()` could take a `Subscription`, and in the function only the fields Recurly accepts for new subscriptions would be sent to ensure nothing breaks (users would just need to look at documentation to know which fields are sent and which are ignored).

Because the structs are all subsets of the `Subscription` struct, this would just require a naming change for library users upgrading but all of the field names would remain the same.

### Support the `context` package
This is the way to handle cancelation, deadlines, etc in Go. In our case we have context propagating across our entire app but not into the Recurly client. I can think of a few ways to go about this:

1) Requiring `context.Context` as the first argument to each method
This is pretty standard at this point. Downside is some would find this a bit tedious to add `context.Background()` to all of the API methods.

2) Adding secondary `Context()` functions for each item.
e.g. `Accounts.Get(id)` and `Accounts.GetContext(ctx, id)`. Internally we would always pass a context around, but `Accounts.Get()` would just be an convenience function that delegates to the context version of the fn. This is similar to what `database/sql` did when they added context support in Go 1.7.

This makes it optional, but it doubles the surface area of the interfaces/mocks. However it's not a lot of code to add.

## Final Notes
While this looks like a lot to update, many of the updates/improvements are internal. There are several minor adjustments that would require users to do some renaming, but I anticipate they will be small changes.

The biggest change this will have on library users is how they handle errors and dealing with  the response type going away in all function calls.

I would appreciate some comments/feedback/suggestions, as well as any other areas we may want to consider revisiting with this breaking change.

**Comment by cristiangraz** [comment]
I'm going to tag a few people I know use the library or have contributed. I know there are many more, and there are others who have contributed who I've missed. Apologies in advance, however because I know this is a critical library for many of you (deals with collecting money!) I opted to tag a few of you to ensure you see this and have an opportunity to contribute thoughts or issues.

@frankienicoletti
@everdance
@qhenkart
@MOZGIII
@ksenish
@laynemcnish
@cfortuner
@bobholt
@bagbee
@jpgnotgif
@nelz9999
@bhelx

**Comment by MOZGIII** [comment]
One other thing I can suggest is using a custom interface instead of `*http.Client`, that is implemented by `*http.Client`.

Here's the code illustrating what I mean:

```go
import "net/http"

type HTTPClient interface {
	Do(req *http.Request) (*http.Response, error)
}

// Enforce compatibility at compile time.
var _ HTTPClient = (*http.Client)(nil)

type Client struct {
	client HTTPClient
	// ...
}
```

**Comment by MOZGIII** [comment]
One more thing: it's going to be an API-breaking change, so the new code should probably have a different import path. This comes from the rules of Go's import path compatibility contract that the lang team encourages people to use. It's up to you, though I'd probably ignore that rule in this particular case - as the API is quite.

Oh, and you might want to take a look at `golang.org/x/xerrors` package. Maybe not immediately for the refactoring, but as a separate step after it - cause it might be too much to change at once.

**Comment by bhelx** [comment]
I'll take a deeper look at it soon. Just want to say 👍 for the transparency in your planning here.

**Comment by bobholt** [comment]
These all sound great to me @cristiangraz. Specifically, I think we would find improved error handling to be very useful.

Specific to handling of `context`, the [Braintree Go library](https://github.com/braintree-go/braintree-go) added it as the first argument to each method. It seemed like a big change, but it turned out to be a pretty straightforward refactor when I applied it to our app last year.

We've abstracted the calls to this library in our app so that we have a limited surface area to change, so no concerns about breaking change from me, if it leads to a stronger library. Thanks for the transparency.

**Comment by nelz9999** [comment]
Sort of following what @MOZGIII, if you're going to breaking API, I might even go further and suggest looking at doing this as a brand new repo, leaving this one with much deprecation warnings. This would even give you a clean Go Modules blank slate to start from too.

But yes, kudos for being transparent and inviting input!

**Comment by ksenish** [comment]
Sounds good to me as well. Thank you for the well-thought-out RFC, explanation and the head's up in general.

Very excited about the upcoming changes, improvements to the error messages and consolidated structs especially!

**Comment by everdance** [comment]
All the changes sound great. thank you for planning these improvements.

Regarding `Reduce extra structs`, there's constraint on `subscription` that prevent us from unifying resource response and request payload as one struct entity. e.g the underlying xml formats are different for `plan`, `account` fields . nevertheless, i think we can at least unify all request payloads as one struct type.

Also i wonder if we should enable users to set the recurly API version instead of hardcoding it right now in client package. As in some cases, users need to stick to previous versions for a while.

**Comment by frankienicoletti** [comment]
It would be less overhead to make a new repo with Go modules. Backwards compatibility won't be an issue and people can migrate over when it's convenient for them. 

If it's not too much of an extra lift to add non-context methods I'd say do it - for example if you could wrap a method and use it in both the context and non-context method. If it's a lot of overhead just do the context methods. That will just be part of the migration to the new library. 

I don't think you should expose the header for the Recurly API version. We have previously made changes to this library along with breaking API changes and allowing those to get out of sync can cause problems. The current versioning method, where a consumer of this library would need to version the dependency module instead of the Recurly API version, is still the safest option. 

Besides that, I think your plan is great 👍!

**Comment by cristiangraz** [comment]
Thanks to everyone for contributing such great feedback!

There's a few comments on creating a new repo in order to avoid breaking changes. While I definitely understand the reasoning and intent, I do not want to maintain multiple repos and have to potentially backport fixes and/or reply to requests. Maintaining this library is time-consuming as is and I would prefer to put all of my efforts into a single version I can really focus on and maintain well.

I will, however, plan to make a detailed migration guide that outlines all of the changes and how to upgrade to make the migration as seamless as possible as well as documenting more examples in the godoc/readme.

@MOZGIII On the interface wrapping `Do(req *http.Request) (*http.Response, error)` vs receiving an `http.Client`: is there a specific limitation or issue you've run into that this would resolve? Since these are all HTTP requests and we provide all of the mocks you'd need to test without making HTTP requests, I'm curious to hear or see example(s) of where this would alleviate a pain point for you so I can better understand.

@everdance
> Also i wonder if we should enable users to set the recurly API version instead of hardcoding it right now in client package. As in some cases, users need to stick to previous versions for a while. -

@frankienicoletti covered my exact thoughts: making the version alone configurable could cause  unintended consequences (that are potentially difficult to debug): it's very possible that the version provided by the user does not match up with what the repo's code is doing.

**Comment by cristiangraz** [comment]
Thanks again for all of the feedback!

I've implemented the majority of the refactor (excluding tests, which I am working on now) and have some updates after implementing the RFC:

## Timeline
Assuming nothing unexpected pops up externally, I plan on having a PR open with a completed refactor by Tuesday or Wednesday for review and early testing by anyone willing to try it out.

## Error handling
No issues here, this is implemented as described above.

## Standardize and simplify pagination
A few things I noticed:
 - `Params` were almost only sent for paginated-requests only
 - `Params` are standardized, with only two requests that I could find sending non-pagination query strings

I opted to:
 - create a pager with a `Fetch` and `FetchAll` methods
 - standardize `recurly.Params` as a struct with an `extra map[string]interface{}` field for occasional custom-values.

```go
// Params are used to send pagination parameters with paginated requests.
type Params struct {
	PerPage   int
	Sort      string
	Order     string
	BeginTime NullTime
	EndTime   NullTime

	State string // supported by some endpoints. Check Recurly's documenation.
	Type  string // supported by some endpoints. Check Recurly's documentation.

	// extra is for any one-off URL params used by a specific endpoint.
	// Values sent as time.Time or recurly.NullTime will be automatically
	// converted to a valid datetime format for Recurly.
	extra map[string]interface{}

	cursor string // managed internally by this library
}
```

Here is an example of usage for accounts:

```go
pager := client.Accounts.List(&recurly.Params{
	PerPage: 50,
	Sort: "asc",
})

// Introduced a new Count() method for all paginated requests
count, err := pager.Count(ctx)
for pager.Next() {
	a, err := pager.Fetch(ctx)
	if err != nil {
		return err
	}
	// a holds []recurly.Account
}
```

Or to have the library fetch all of the results automatically:

```go
// Replace nil with &recurly.Params{} if desired
a, err := client.Accounts.List(nil).FetchAll(ctx)
```

## Reduce extra structs
The idea was to use a single `Subscription` struct, and internally shift fields around so we could avoid `NewSubscription` and `UpdateSubscription` types.

After diving in this ended up being unfeasible without introducing a lot of complexity and ultimately confusion for users.

## Support the `context` package
Based on the comments and several other libraries I opted for `context.Context` as the first argument to each method. This is now fairly standard and an easy change.

**Comment by MOZGIII** [comment]
> @MOZGIII On the interface wrapping `Do(req *http.Request) (*http.Response, error)` vs receiving an `http.Client`: is there a specific limitation or issue you've run into that this would resolve? Since these are all HTTP requests and we provide all of the mocks you'd need to test without making HTTP requests, I'm curious to hear or see example(s) of where this would alleviate a pain point for you so I can better understand.

I don't have a particular pain point with it. However, I consider it a good move for the following reasons:
- it follow the go best practice of accepting non-concrete values (an returning concrete values);
- it explicitly narrows the interface the code relies on;
- it allows one to pass some other implementation that is not the standard `*http.Client`; this might be useful to someone using `https://github.com/gojektech/heimdall/httpclient` for instance

**Comment by cristiangraz** [comment]
I've opened up a PR with this change and created a migration guide outlining all of the steps needed to migrate.

I'm going to finish up documentation, then switch over our staging and production code to pull the new `refactor` branch. If anyone here is able to try out the new PR and provide feedback/report bugs, I would greatly appreciate it!

Please note that there may be some minor breaking changes here-and-there while the PR is readied for production use, however I expect very minimal changes as the client API is pretty solidified at this point.

https://github.com/blacklightcms/recurly/pull/118
https://github.com/blacklightcms/recurly/wiki/v1-Migration-Guide

**Comment by cristiangraz** [comment]
@MOZGIII I've changed `*http.Client` to the new `HTTPDoer` interface.

**Comment by cristiangraz** [comment]
PR for this is open (#118) and migration guide is complete. I've also switched our production code to the new branch. Unfortunately I'm still waiting for a response from Recurly support on the purchases API.

Over the next few days I'll be adding additional documentation, examples, and testing examples while I wait. If anyone here is able to try the `refactor` branch I'd really appreciate it! It dramatically simplified our implementation.

**Comment by cristiangraz** [comment]
This has been released as `v1.0.0`
